xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiAction.java

```diff
@@ -731,45 +732,30 @@ public void execute(XWikiContext context) throws Exception
     }
 
     /**
-     * Check if the given document exists or not and if it should return a 404 based on the context.
-     * If the action is view, then it should returns a 404 for all viewers except recyclebin, children and siblings.
-     * If the action is get, it should returns a 404 if the sheet parameter is not used (or if the given sheet doesn't
-     * exist either) and if the disableCheckNotExisting parameter is not used.
+     * Check if the given document exists or not and if it should return a 404 based on the context. A {@link
+     * DocExistValidator} with an hint matching the current action is used to check if the document exists. When no
+     * {@link DocExistValidator} is found, the response is always {@code false} When a {@link DocExistValidator} is
+     * found, the result is delegated to {@link DocExistValidator#docExist(XWikiDocument, XWikiContext)}.
      *
-     * @param doc the doc for which to check it exists or not.
+     * @param doc the doc for which to check it exists or not
      * @param context the current context
      * @return {@code true} if we should return a 404
      */
     private boolean shouldReturnDocDoesNotExist(XWikiDocument doc, XWikiContext context)
     {
         boolean result = false;
-        if (doc.isNew()) {
-            String action = context.getAction();
-            XWikiRequest request = context.getRequest();
-            if ("view".equals(action)) {
-                result = !"recyclebin".equals(request.get("viewer"))
-                    && !"children".equals(request.get("viewer"))
-                    && !"siblings".equals(request.get("viewer"));
-            } else if ("get".equals(action)) {
-                String sheet = request.get("sheet");
-                if (!StringUtils.isEmpty(sheet)) {
-                    DocumentReference sheetReference = getCurrentMixedDocumentReferenceResolver().resolve(sheet);
-                    try {
-                        XWikiDocument sheetDoc = context.getWiki().getDocument(sheetReference, context);
-                        result = sheetDoc.isNew();
-                    } catch (XWikiException e) {
-                        LOGGER.warn("Error while trying to load sheet [{}] for checking status code "
-                            + "on GET request for [{}]: [{}]", sheetReference, doc.getDocumentReference(),
-                            ExceptionUtils.getRootCauseMessage(e));
-                        // there is an error we consider that the sheet doesn't exist.
-                        result = true;
-                    }
-                } else {
-                    result = !"1".equals(request.get("disableCheckNotExisting"));
-                }
+        String action = context.getAction();
+        try {
+            DocExistValidator instance = this.componentManager.getInstance(DocExistValidator.class, action);
```

Comment:
Throwing exceptions is slow and for all actions other than "view" and "get" you will get an exception so would be better to first call this.componentManager.hasComponent to see if there is actually a component available for this action.

Reply:
> This also allows differentiating between "there is no DocExistValidator for this action" and "there is one, but it's crashing, so I should be a bit stricter than a debug log".

Note that I also wondered about that and I double checked the interface documentation of ComponentManager it only says that `ComponentLookupException` is thrown if the component is not found. Not when the component fails to initialize. The doc should probably be improved on this aspect.

Reply:
> The doc should probably be improved on this aspect.

Indeed, that javadoc is wrong. There is an exception if anything prevent to get the component instance.